[css-transitions-2] Add transition-behavior property

josepharhar opened the pull request:
Fixes https://github.com/w3c/csswg-drafts/issues/8857

dbaron commented on css-transitions-2/Overview.bs:
```diff
@@ -120,6 +122,42 @@ the reversing shortening factor.
+
+<pre class="propdef">
+  Name: transition-behavior
+  Value: normal | allow-discrete
```
This needs to allow a list of values, not just a single value.  To do that you should probably make this `<transition-behavior-value>#`.  See https://drafts.csswg.org/css-backgrounds-3/#background-attachment for an example of what this should look like (and probably example bikeshed syntax too).

josepharhar replied:
done, thanks!

dbaron commented on css-transitions-2/Overview.bs:
```diff
@@ -120,6 +122,42 @@ the reversing shortening factor.
+The syntax for specifying 'transition-behavior' is as follows:
+
+<div class="prod">
+  <dfn type id="transition-behavior-value">&lt;transition-behavior-value&gt;</dfn> = [ 'normal' | 'allow-discrete' ]
```
This could be a bit simpler, see the source of css-backgrounds-3 for an example.  Though I admit it matches the complexity in css-transitions-1...

dbaron commented on css-transitions-2/Overview.bs:
```diff
@@ -88,7 +88,9 @@ In CSS Transitions Level 2,
 <em>unless</em> they have an [=animation type=] that is [=not animatable=].
 Values with a [=discrete=] [=animation type=] <em>are</em> [=transitionable=],
-and flip at 50% progress (<var>p</var> = 0.5).
+and flip at 50% progress (<var>p</var> = 0.5) if the 'transition-behavior' is
```
To match the terminology for the correct item in the list in css-transforms-1, this should probably refer to the "matching transition behavior".

flackr replied:
It might be simpler to leave out the "and flip at 50% progress" as this is already defined in web-animations-1. The important part is whether they are transitionable.

If you look at the [definition from css-transitions-1](https://www.w3.org/TR/css-transitions-1/#transitionable):
> When comparing the [before-change style](https://www.w3.org/TR/css-transitions-1/#before-change-style) and [after-change style](https://www.w3.org/TR/css-transitions-1/#after-change-style) for a given property, the property values are transitionable if they have an [animation type](https://drafts.csswg.org/web-animations-1/#animation-type) that is neither [not animatable](https://drafts.csswg.org/web-animations-1/#not-animatable) nor [discrete](https://drafts.csswg.org/web-animations-1/#discrete).

We could update this to something like:

When comparing the [before-change style](https://www.w3.org/TR/css-transitions-1/#before-change-style) and [after-change style](https://www.w3.org/TR/css-transitions-1/#after-change-style) for a given property, the property values are transitionable if:
- They have an [animation type](https://drafts.csswg.org/web-animations-1/#animation-type) that is neither [not animatable](https://drafts.csswg.org/web-animations-1/#not-animatable) nor [discrete](https://drafts.csswg.org/web-animations-1/#discrete), or
- The transition-behavior is allow-discrete and they have an [animation type](https://drafts.csswg.org/web-animations-1/#animation-type) that is [discrete](https://drafts.csswg.org/web-animations-1/#discrete).

josepharhar replied:
Thanks rob! I replaced this paragraph with your text

dbaron commented on css-transitions-2/Overview.bs:
```diff
@@ -88,7 +88,9 @@ In CSS Transitions Level 2,
 property values are [=transitionable=]
 <em>unless</em> they have an [=animation type=] that is [=not animatable=].
```
I think this paragraph should be rewritten a bit more fully to reflect the new approach.

josepharhar replied:
I rewrote the whole thing as per rob's suggestion

dbaron commented on css-transitions-2/Overview.bs:
```diff
@@ -120,6 +122,42 @@ the reversing shortening factor.
+The syntax for specifying 'transition' is as follows:
+
+<div class="prod">
+  <dfn type id="single-transition">&lt;single-transition&gt;</dfn> = <<transition-behavior-value>> || [ ''transition-property/none'' | <<single-transition-property>> ] || <<time>> || <<easing-function>> || <<time>>
```
I think it's probably best to put the `<<transition-behavior-value>>` last; the order does affect the order of serialization that the spec requires.

flackr replied:
We might want to leave the normal value out of the shorthand - it is the initial value so unnecessary to be specified, and it would prevents its specification in other longhands in the future.

dbaron replied:
I think leaving `normal` out of the shorthand is an interesting idea -- but I also think it's not something the WG has done before for any shorthand, so if we think it's worth doing, I think we should bring it up explicitly to the WG to see if other folks have feedback on the idea.

dbaron replied:
I realize that `transition-property` takes `<custom-ident>` values, which means that the spec should probably also have wording excluding the two new keywords from that `<custom-ident>` as described in https://drafts.csswg.org/css-values-4/#identifier-value , although that change does have (low, I think) compatibility risk.

Or, alternatively, we could choose *not* to follow that advice and exclude the keywords, in which case we probably *do* want to leave this in the first position.

josepharhar replied:
> I think it's probably best to put the <<transition-behavior-value>> last; the order does affect the order of serialization that the spec requires.

So instead of doing `transition: allow-discrete display 2s` it should be `transition: display 2s allow-discrete`? Now that the value is called "allow-discrete" I'm more ok with this. When it was just "discrete" or "animatable" I felt more strongly that it should come first. What do you think rob?

> the spec should probably also have wording excluding the two new keywords from that `<custom-ident>`

If we put the keyword at the end, then this wouldn't be needed right?

dbaron replied:
I don't think the ordering is relevant.  This affects the parsing rules -- and effectively without that exclusion the spec would need prose rules for how to distinguish the values, given that https://drafts.csswg.org/css-values-4/#component-combinators explicitly says that the ordering around `||` does *not* matter.  (This implies, I think, that it doesn't imply any rule about breaking ties during parsing when more than one option of an `||` matches.  Though maybe that implication wasn't intended.  But I think it does imply that it's a specification error if more than one option around an `||` matches and the spec doesn't say how to resolve the conflict.)

josepharhar replied:
> given that https://drafts.csswg.org/css-values-4/#component-combinators explicitly says that the ordering around || does not matter

Oh wow, so `transition: 2s color` works the same as `transition: color 2s`!? It looks like this is true after I tried a basic test in chrome.

I moved it to the end of the list

dbaron commented on css-transitions-2/Overview.bs:
```diff
@@ -120,6 +124,40 @@ the reversing shortening factor.
+
+## The 'transition' Shorthand Property ## {#transition-shorthand-property}
+
+The syntax for specifying 'transition' is as follows:
```
```suggestion
The syntax for specifying an item in the 'transition' shorthand is as follows:
```

dbaron commented on css-transitions-2/Overview.bs:
```diff
@@ -120,6 +124,40 @@ the reversing shortening factor.
+
+<div class=prod><dfn><<transition-behavior-value>></dfn> = normal | allow-discrete</pre>
+
+When ''transition-behavior-value/normal'' is specified, animations will not be
```
```suggestion
When ''transition-behavior-value/normal'' is specified, transitions will not be
```

dbaron commented on css-transitions-2/Overview.bs:
```diff
@@ -120,6 +124,40 @@ the reversing shortening factor.
+
+When ''transition-behavior-value/normal'' is specified, animations will not be
+started for discrete properties, only for interpolable properties. When
+''allow-discrete'' is specified, animations will be started for discrete
```
```suggestion
''allow-discrete'' is specified, transitions will be started for discrete
```

dbaron commented on css-transitions-2/Overview.bs:
```diff
@@ -84,11 +84,15 @@ the transition was generated (e.g. 'border-left-width').
-In CSS Transitions Level 2,
-property values are [=transitionable=]
-<em>unless</em> they have an [=animation type=] that is [=not animatable=].
-Values with a [=discrete=] [=animation type=] <em>are</em> [=transitionable=],
-and flip at 50% progress (<var>p</var> = 0.5).
```
I think it's worth adding this back as a note.  In bikeshed that should be doable by making it its own paragraph beginning with "Note: ", though I think in full it probably needs to be something like: "Note: When values with a [=discrete=] [=animation type] are transitioned, they flip at 50% progress."

josepharhar replied:
done, thanks!

dbaron reviewed (commented):
This basically looks good, just a few comments:

josepharhar commented:
In my last commit ce5d6d3 I removed a note about transition:all which is no longer true based on the way I implemented the new CSS property: https://chromium-review.googlesource.com/c/chromium/src/+/4616155/14/third_party/blink/web_tests/external/wpt/css/css-transitions/all-with-discrete.tentative.html

@dbaron do you think this is ok?

dbaron reviewed (approved):
Good catch.  I agree that `all` combined with `allow-discrete` should actually allow discrete transitions!
